File: julia/mandelbrot-set/mandelbrot_set.jl
import Base.+, Base.*, Base.show, Base.abs

abstract type AbstractComplex end

struct MyComplex{T} <: AbstractComplex
	real_part::T
	imag_part::T
end

function +(z1::AbstractComplex, z2::AbstractComplex)
	MyComplex(real(z1)+real(z2), imag(z1) + imag(z2))
end

function *(x::AbstractComplex, y::AbstractComplex)
	r = real(x)*real(y) - imag(x)*imag(y)
	i = real(x)*imag(y) + imag(x)*real(y)
	MyComplex(r, i)
end

real(z::AbstractComplex) = z.real_part
imag(z::AbstractComplex) = z.imag_part
abs(z::AbstractComplex) = sqrt(real(z)^2 + imag(z)^2)

function show(io::IO,z::AbstractComplex)
	sign = imag(z) >= 0 ? "+" : "-"
	println("$(real(z)) $sign j $(abs(imag(z)))")
end

function findMandelbrotColor(c::AbstractComplex; N::Int=60, absmax=1e4)
	n = 0
	z = MyComplex(0., 0.)
	while abs(z) <= absmax && n < N
		z += z*z + c
		n += 1
	end
	return 1 - n / N
end


using Plots
xAxis, yAxis = -2:0.001:0.3, -1:0.005:1
z = Float64[]

for (x, y) in Iterators.product(xAxis, yAxis)
	push!(z, findMandelbrotColor(MyComplex(x, y)))
end

heatmap(xAxis, yAxis, z, legend=false, xaxis=false, yaxis=false)

File: python/mandelbrot-set/mandelbrot_set.py
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import itertools as iter


class MyComplex:

    def __init__(self, real_part, imag_part):
        self.real_part = real_part
        self.imag_part = imag_part

    def __add__(self, z):
        return MyComplex(real(self) + real(z), imag(self) + imag(z))

    def __mul__(self, z):
        real_part = real(self) * real(z) - imag(self) * imag(z)
        imag_part = real(self) * imag(z) + imag(self) * real(z)
        return MyComplex(real_part, imag_part)

    def __repr__(self):
        sign = '+' if self.imag_part >= 0 else '-'
        return f"{self.real_part} {sign} j {abs(self.imag_part)}"


def real(z: MyComplex):
    return z.real_part


def imag(z: MyComplex):
    return z.imag_part


def magnitude(z: MyComplex):
    return (real(z)**2 + imag(z)**2)**0.5


def findMandelbrotColor(c: MyComplex, N=60, absmax=1e4):
    n = 0
    z = MyComplex(0, 0)
    while magnitude(z) <= absmax and n < N:
        z = z + z * z + c
        n = n + 1
    return 1 - n / N


xAxis, yAxis = np.arange(-2, 0.3, 0.01), np.arange(-1, 1, 0.005)
z = []
for (x_, y_) in iter.product(xAxis, yAxis):
    z.append(findMandelbrotColor(MyComplex(x_, y_)))

z = np.reshape(z, (len(xAxis), len(yAxis)))
z = np.transpose(z)
plt.imshow(z, origin='lower', extent=[
           xAxis.min(), xAxis.max(), yAxis.min(), yAxis.max()])
plt.show()
